atakama/rule_engine.py

```diff
@@ -0,0 +1,195 @@
+
+
+@dataclass
+class ApprovalRequest:
+    request_type: RequestType
+    """Request type"""
+    device_id: bytes
+    """Requesting device uuid"""
+    profile: ProfileInfo
+    """"""
+    auth_meta: List[MetaInfo]
+    """Authenticated metadata associated with the encrypted data.   Typically a path to a file."""
+
+
+class RulePlugin(Plugin):
+    """
+    Base class for key server approval rule handlers.
+
+    When a key server receives a request, rules are consulted for approval.
+
+    Each rule receives its configuration from the policy file,
+    not the atakama config, like other plugins.
+    """
+
+    @abc.abstractmethod
+    def approve_request(self, request: ApprovalRequest) -> bool:
+        """
+        Return True if the request to decrypt a file will be authorized.
+
+        Return False if the request is to be denied.
+        Raise None if the request type is unknown or invalid.
+        Exceptions and None are logged, and considered False.
+
+        This is called any time:
+            a decryption agent wishes to decrypt a file.
+            a decryption agent wishes to search a file.
+            a decryption agent wishes to create a new multifactor profile.
+
+        See the RuleRequest class for more information
+        """
+
+    def check_quota(self, profile: ProfileInfo) -> bool:
+        """
+        Returns False if the profile will not be approved in the next request.
+        Returns True if the profile *may* be approved for access, and is not past a limit.
+
+        This is not a guarantee of future approval, it's a way of checking to see if any users have
+        reached any limits, quotas or other stateful things for reporting purposed.
+        """
+
+    def clear_quota(self, profile: ProfileInfo) -> None:
+        """
+        Reset or clear any limits, quotas, access counts, bytes-transferred for a given profile.
+
+        Used by an administrator to "clear" or "reset" a user that has hit limits.
+        """
+
+    @classmethod
+    def from_dict(cls, data: dict) -> "RulePlugin":
+        """
+        Factory function called with a dict from the rules yaml file.
+        """
+
+        assert type(data) is dict, "Rule entries must be dicts"
+        assert "plugin" in data, "Rule entries must have a plugin name"
+        pname = data.pop("plugin")
+        p = RulePlugin.get_by_name(pname)(data)
+        assert isinstance(p, RulePlugin), "Rule plugins must derive from RulePlugin"
+        return p
+
+
+class RuleSet(List[RulePlugin]):
+    """A list of rules, can reply True, False, or None to an ApprovalRequest"""
+
+    def approve_request(self, request: ApprovalRequest) -> bool:
+        for rule in self:  # pylint: disable=not-an-iterable
+            try:
+                res = rule.approve_request(request)
+                if res is None:
+                    log.error("unknown request type error in rule %s", rule)
+                return bool(res)
```

Comment:
bool conversion should be unnecessary right?

Reply:
if res is None, then we convert to bool, because, for now, "None" is not actionable - except for logging.  hence this returns bool only

Reply:
right I see now